pallets/loans/src/interest.rs

```diff
@@ -0,0 +1,119 @@
+            let util = Self::calc_utilization_ratio(total_cash, total_borrows, total_reserves)?;
+
+            let borrow_rate = market
+                .rate_model
+                .get_borrow_rate(util)
+                .ok_or(ArithmeticError::Overflow)?;
+            let supply_rate =
+                InterestRateModel::get_supply_rate(borrow_rate, util, market.reserve_factor);
+
+            UtilizationRatio::<T>::insert(currency_id, util);
+            BorrowRate::<T>::insert(currency_id, &borrow_rate);
+            SupplyRate::<T>::insert(currency_id, supply_rate);
+
+            Self::update_borrow_index(borrow_rate, currency_id, &market)?;
+            Self::update_exchange_rate(currency_id)?;
+        }
+
+        Ok(())
+    }
+
+    /// Update the exchange rate according to the totalCash, totalBorrows and totalSupply.
+    ///
+    /// exchangeRate = (totalCash + totalBorrows - totalReserves) / totalSupply
+    pub(crate) fn update_exchange_rate(currency_id: CurrencyId) -> DispatchResult {
+        let total_supply = Self::total_supply(currency_id);
+        if total_supply.is_zero() {
+            return Ok(());
+        }
+        let total_cash = Self::get_total_cash(currency_id);
+        let total_borrows = Self::total_borrows(currency_id);
+        let total_reserves = Self::total_reserves(currency_id);
+
+        let cash_plus_borrows_minus_reserves = total_cash
+            .checked_add(total_borrows)
+            .and_then(|r| r.checked_sub(total_reserves))
+            .ok_or(ArithmeticError::Overflow)?;
+        let exchange_rate =
+            Rate::checked_from_rational(cash_plus_borrows_minus_reserves, total_supply)
+                .ok_or(ArithmeticError::Underflow)?;
+
+        ExchangeRate::<T>::insert(currency_id, exchange_rate);
+
+        Ok(())
+    }
+
+    /// Calculate the borrowing utilization ratio of the specified market
+    ///
+    /// utilizationRatio = totalBorrows / (totalCash + totalBorrows − totalReserves)
+    pub(crate) fn calc_utilization_ratio(
+        cash: Balance,
+        borrows: Balance,
+        reserves: Balance,
+    ) -> Result<Ratio, DispatchError> {
+        // utilization ratio is 0 when there are no borrows
+        if borrows.is_zero() {
+            return Ok(Ratio::zero());
+        }
+        let total = cash
+            .checked_add(borrows)
+            .and_then(|r| r.checked_sub(reserves))
+            .ok_or(ArithmeticError::Overflow)?;
+
+        Ok(Ratio::from_rational(borrows, total))
+    }
+
+    /// Update the borrow index by borrow rate, the total borrows and
+    /// total reserves will be updated simultaneously.
+    ///
+    /// interestAccumulated = totalBorrows * borrowRate
+    /// totalBorrows = interestAccumulated + totalBorrows
+    /// totalReserves = interestAccumulated * reserveFactor + totalReserves
+    /// borrowIndex = borrowIndex * (1 + borrowRate)
+    pub(crate) fn update_borrow_index(
```

Comment:
maybe we can add https://substrate.dev/rustdocs/latest/frame_support/attr.require_transactional.html marker so that we are sure this function will only be used in transactional calls

Reply:
It will break the test because the function is not called within a storage transaction.
It will cause panic: 'Require transaction not called within with_transaction'.